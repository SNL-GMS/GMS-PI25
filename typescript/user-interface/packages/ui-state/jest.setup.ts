/* eslint-disable max-classes-per-file */
/* eslint-disable @typescript-eslint/no-unused-vars */
/* eslint-disable import/first */
/* eslint-disable import/no-extraneous-dependencies */
import 'fake-indexeddb/auto';

import type { FilterDefinition } from '@gms/common-model/lib/filter/types';
import crypto from 'crypto';
import * as util from 'util';

// This is to fix an issue with test leaks related to createStateSyncMiddleware() in jsdom
process.env.GMS_DISABLE_REDUX_STATE_SYNC = 'true';

Object.defineProperty(global, 'crypto', {
  value: {
    subtle: {
      digest: async (algorithm: AlgorithmIdentifier, msg: Uint8Array) =>
        (crypto.webcrypto as any).subtle.digest(algorithm, msg) as Promise<any>
    }
  }
});

Object.defineProperty(window, 'TextEncoder', {
  writable: true,
  value: util.TextEncoder
});
Object.defineProperty(window, 'TextDecoder', {
  writable: true,
  value: util.TextDecoder
});
Object.defineProperty(global, 'TextEncoder', {
  writable: true,
  value: util.TextEncoder
});
Object.defineProperty(global, 'TextDecoder', {
  writable: true,
  value: util.TextDecoder
});

import Adapter from '@cfaester/enzyme-adapter-react-18';
import Enzyme, { mount, render, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import type { GlobalWithFetchMock } from 'jest-fetch-mock';

const customGlobal: GlobalWithFetchMock = (global as unknown) as GlobalWithFetchMock;
// eslint-disable-next-line @typescript-eslint/no-var-requires, @typescript-eslint/no-require-imports
customGlobal.fetch = require('jest-fetch-mock');

customGlobal.fetchMock = customGlobal.fetch;

// eslint-disable-next-line @typescript-eslint/no-var-requires, @typescript-eslint/no-require-imports
require('jest-canvas-mock');

Enzyme.configure({ adapter: new Adapter() });

const globalAny: any = global;

// eslint-disable-next-line @typescript-eslint/no-var-requires, @typescript-eslint/no-require-imports
globalAny.fetch = require('jest-fetch-mock');

// Make Enzyme functions available in all test files without importing
globalAny.shallow = shallow;
globalAny.render = render;
globalAny.mount = mount;
globalAny.toJson = toJson;
globalAny.TextEncoder = util.TextEncoder;

globalAny.window = window;

class Worker {
  public url: string;

  public onmessage: (message: Record<string, unknown>, transfer?: Transferable[]) => void;

  public constructor(stringUrl: string) {
    this.url = stringUrl;
    this.onmessage = () => {
      // do nothing
    };
  }

  public postMessage(message: Record<string, unknown>, transfer?: Transferable[]) {
    this.onmessage(message, transfer);
  }
}
globalAny.window.Worker = Worker;

class MessagePort {
  public url: string;

  public onmessage: (message: Record<string, unknown>, transfer?: Transferable[]) => void;

  public postMessage(message: Record<string, unknown>, transfer?: Transferable[]) {
    this.onmessage(message, transfer);
  }
}

class SharedWorker {
  public url: string;

  public port = new MessagePort();

  public constructor(stringUrl: string) {
    this.url = stringUrl;

    this.port.onmessage = () => {
      // do nothing
    };
  }
}
globalAny.window.SharedWorker = SharedWorker;

// eslint-disable-next-line @typescript-eslint/no-var-requires, @typescript-eslint/no-require-imports,  import/no-extraneous-dependencies
require('jest-fetch-mock').enableMocks();

jest.mock('@gms/ui-wasm', () => {
  const isDesigned = jest.requireActual('@gms/ui-wasm/lib/gms-filters-provider/is-designed');

  const filter = async (
    filterDefinition: FilterDefinition,
    data: Float64Array
  ): Promise<Float64Array> => Promise.resolve(data);

  const design = async (filterDefinition: FilterDefinition): Promise<FilterDefinition> =>
    Promise.resolve(filterDefinition);

  const getBoundsForPositionBuffer = async (): Promise<{
    min: number;
    minSecs: number;
    max: number;
    maxSecs: number;
  }> => Promise.resolve({ min: 0, minSecs: 0, max: 10, maxSecs: 10 });

  return {
    design,
    filter,
    getBoundsForPositionBuffer,
    isDesigned: isDesigned.isDesigned
  };
});
