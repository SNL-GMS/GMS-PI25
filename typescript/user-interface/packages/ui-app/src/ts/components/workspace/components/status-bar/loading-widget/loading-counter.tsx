import { selectRequestStats, useAppSelector } from '@gms/ui-state';
import React from 'react';

/**
 * Simple component to render the count
 */
export const LoadingCounter = React.memo(function LoadingCounter() {
  const { initiated, completed } = useAppSelector(selectRequestStats);
  return <span className="loading-counter">{`${completed}/${initiated}`}</span>;
});
