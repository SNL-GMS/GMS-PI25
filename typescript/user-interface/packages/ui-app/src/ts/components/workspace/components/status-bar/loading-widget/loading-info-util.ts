import { IconNames } from '@blueprintjs/icons';
import { getFriendlyNames } from '@gms/ui-state';
import type { RequestStatus } from '@gms/ui-state/lib/app/state/analyst/types';
import memoizeOne from 'memoize-one';

const memoizedGetFriendlyNames = memoizeOne(getFriendlyNames);

/**
 * Get a human readable (friendly) name from the defined friendly names per request, if the url
 * provided matches any of the requests that have a friendly name defined.
 *
 * If no friendly name is found, returns the path from the URL.
 *
 * @param url A url to attempt to match to a friendly name
 * @returns a human readable friendly name, or url path
 */
export const getFriendlyNameFromRequest = (url: string): string | undefined => {
  if (!url) return undefined;
  const friendlyNames = memoizedGetFriendlyNames();
  if (friendlyNames[url]) return friendlyNames[url];
  const match = url?.match(/^https?:\/\/[A-Za-z0-9:.]*(\/.*\/?)$/);
  return match?.[1]; // index 1 is the matched substring
};

/**
 * String representing the error icon name
 */
export const ErrorIcon = IconNames.ISSUE;

/**
 * String representing the completed icon name
 */
export const CompleteIcon = IconNames.TICK;

/**
 * Returns the icon name for error or completed requests, or undefined.
 */
export function getIconName(requestStatus: RequestStatus) {
  if (requestStatus.error) {
    return ErrorIcon;
  }
  if (requestStatus.isComplete) {
    return CompleteIcon;
  }
  return undefined;
}
