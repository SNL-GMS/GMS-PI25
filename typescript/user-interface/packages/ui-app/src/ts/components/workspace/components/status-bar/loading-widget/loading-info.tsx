import { selectIsLoading, selectPendingRequests, useAppSelector } from '@gms/ui-state';
import React from 'react';

import { getFriendlyNameFromRequest } from './loading-info-util';

const LoadingCompleteDescription = 'Complete';

/**
 * Displays the loading text for the LoadingWidget, which is either the last of the pending requests,
 * or else a "Loading: Complete" indicator.
 */
export const LoadingInfo = React.memo(function LoadingInfo() {
  const pendingRequests = Object.values(useAppSelector(selectPendingRequests));
  const loadingText = getFriendlyNameFromRequest(
    pendingRequests?.[pendingRequests.length - 1]?.url
  );
  const isLoading = useAppSelector(selectIsLoading);
  return (
    <div className="loading-info">
      <span className="loading-info__label">Loading: </span>
      {isLoading ? loadingText ?? LoadingCompleteDescription : LoadingCompleteDescription}
    </div>
  );
});
