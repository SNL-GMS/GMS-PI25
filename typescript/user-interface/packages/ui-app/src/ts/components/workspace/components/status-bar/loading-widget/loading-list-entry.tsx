import { Icon, Spinner } from '@blueprintjs/core';
import type { RequestStatus } from '@gms/ui-state/lib/app/state/analyst/types';
import classNames from 'classnames';
import React from 'react';

import { getFriendlyNameFromRequest, getIconName } from './loading-info-util';

/**
 * Returns an error message, or null if the error is undefined.
 *
 * @param error the error from which to get the message. Errors have `any` type, so we use that it here, too
 * @returns an error message string, or null
 */
const getErrorMessage = (error: any): string | null => {
  if (!error) {
    return null;
  }
  if (typeof error === 'string') {
    return error;
  }
  if (error?.message) {
    return error.message;
  }
  return 'See developer console for more details';
};

/**
 * The type of the props for the {@link LoadingListEntry} component
 */
export interface LoadingListEntryProps {
  requestStatus: RequestStatus;
}

/**
 * Creates a single loading list entry for the ordered list in the LoadingInfoTooltip
 */
export const LoadingListEntry = React.memo(function LoadingListEntry({
  requestStatus
}: LoadingListEntryProps) {
  const iconName = getIconName(requestStatus);
  return (
    <li
      className={classNames({
        'loading-list-entry': true,
        'is-error': !!requestStatus.error,
        'is-complete': requestStatus.isComplete,
        'is-pending': !requestStatus.isComplete
      })}
    >
      <div className="loading-list-entry__description">
        {!requestStatus.isComplete ? (
          <Spinner className="loading-list-entry__spinner" size={14} intent="none" />
        ) : (
          <Icon className="loading-list-entry__icon" icon={iconName} />
        )}
        {requestStatus.error ? <span className="loading-list-entry__label">Error: </span> : null}
        <span className="loading-list-entry__value">
          {`${getFriendlyNameFromRequest(requestStatus.url)}`}
        </span>
      </div>
      {requestStatus.error ? (
        <span className="loading-list-entry__more-info">
          {getErrorMessage(requestStatus.error)}
        </span>
      ) : null}
    </li>
  );
});
