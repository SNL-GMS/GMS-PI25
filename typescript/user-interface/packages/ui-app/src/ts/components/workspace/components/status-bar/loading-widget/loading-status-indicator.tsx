import { Icon, Spinner } from '@blueprintjs/core';
import { selectHasRequestErrorOccurred, selectIsLoading, useAppSelector } from '@gms/ui-state';
import React from 'react';

import { CompleteIcon, ErrorIcon } from './loading-info-util';

/**
 * Creates an icon or an indeterminate spinner to indicate the status of the app's requests.
 * Will be a spinner if loading, then an error icon if an error occurred, then a tick if complete.
 */
export const LoadingStatusIndicator = React.memo(function LoadingStatusIndicator() {
  const isLoading = useAppSelector(selectIsLoading);
  const isError = useAppSelector(selectHasRequestErrorOccurred);
  if (isLoading) {
    return <Spinner className="loading-widget__spinner" size={20} intent="none" />;
  }
  if (isError) {
    return <Icon icon={ErrorIcon} size={16} />;
  }
  return <Icon icon={CompleteIcon} size={20} />;
});
