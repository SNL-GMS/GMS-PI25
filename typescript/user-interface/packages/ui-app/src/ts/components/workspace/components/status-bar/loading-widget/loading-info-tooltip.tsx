import { selectRequestStats, selectRequestStatuses, useAppSelector } from '@gms/ui-state';
import classNames from 'classnames';
import React from 'react';

import { LoadingListEntry } from './loading-list-entry';

const MAX_NUM_REQUESTS_TO_RENDER = 500;

/**
 * Creates the contents of the loading info tooltip: a list of each entry that is being loaded.
 */
export const LoadingInfoTooltip = React.memo(function LoadingInfoTooltip() {
  const requests = useAppSelector(selectRequestStatuses);
  const { initiated, completed } = useAppSelector(selectRequestStats);
  const numRequests = Object.values(requests).length;
  const listEntries = React.useMemo(
    () =>
      Object.entries(requests)
        .slice(-1 * Math.min(numRequests, MAX_NUM_REQUESTS_TO_RENDER))
        .reverse()
        .map(([id, requestStatus]) => {
          return <LoadingListEntry key={id} requestStatus={requestStatus} />;
        }),
    [numRequests, requests]
  );
  return (
    <div>
      <div className={classNames(['loading-list__sticky-label'])}>
        {`Loaded ${completed} out of ${initiated} requests`}
      </div>
      <ol className="loading-list">
        {listEntries}
        {numRequests > MAX_NUM_REQUESTS_TO_RENDER ? (
          <li className="loading-list-entry__description loading-list-entry__description--set-apart">
            . . . {numRequests - MAX_NUM_REQUESTS_TO_RENDER} older requests not shown . . .
          </li>
        ) : null}
      </ol>
    </div>
  );
});
