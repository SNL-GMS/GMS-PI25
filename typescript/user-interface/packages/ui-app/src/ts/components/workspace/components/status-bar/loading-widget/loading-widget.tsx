import { Classes, PopoverPosition } from '@blueprintjs/core';
import { Classes as PopoverClasses, Popover2 } from '@blueprintjs/popover2';
import classNames from 'classnames';
import React from 'react';

import { LoadingCounter } from './loading-counter';
import { LoadingInfo } from './loading-info';
import { LoadingInfoTooltip } from './loading-info-tooltip';
import { LoadingStatusIndicator } from './loading-status-indicator';

/**
 * Creates a loading widget for the status bar that shows a loading spinner, the number of requests complete,
 * the number in flight, and a description of the latest request made.
 */
export const LoadingWidget = React.memo(function LoadingWidget() {
  return (
    <Popover2
      content={<LoadingInfoTooltip />}
      popoverClassName={classNames([PopoverClasses.TOOLTIP2, Classes.DARK])}
      hoverOpenDelay={300}
      position={PopoverPosition.TOP_LEFT}
      autoFocus={false}
      enforceFocus={false}
      canEscapeKeyClose
      minimal
    >
      <div className="loading-widget" role="button">
        <LoadingStatusIndicator />
        <LoadingCounter />
        <LoadingInfo />
      </div>
    </Popover2>
  );
});
