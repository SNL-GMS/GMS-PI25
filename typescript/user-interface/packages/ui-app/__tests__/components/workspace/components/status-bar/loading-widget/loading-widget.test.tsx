import '@testing-library/jest-dom';

import { analystActions, getStore } from '@gms/ui-state';
import type { RequestTrackerMessage } from '@gms/ui-workers';
import { cleanup, fireEvent, render, screen } from '@testing-library/react';
import React from 'react';
import { Provider } from 'react-redux';

import { LoadingWidget } from '~components/workspace/components/status-bar/loading-widget/loading-widget';

describe('LoadingWidget', () => {
  const store = getStore();
  // set up request tracking redux state
  const pendingRequestTrackers: RequestTrackerMessage[] = [
    {
      id: '1',
      message: 'REQUEST_INITIATED',
      url: 'https://example.com/foo'
    },
    {
      id: '2',
      message: 'REQUEST_INITIATED',
      url: 'https://example.com/bar'
    },
    {
      id: '3',
      message: 'REQUEST_INITIATED',
      url: 'https://example.com/baz'
    },
    {
      id: '4',
      message: 'REQUEST_INITIATED',
      url: 'https://example.com/foo'
    },
    {
      id: '5',
      message: 'REQUEST_INITIATED',
      url: 'https://example.com/bar'
    },
    {
      id: '6',
      message: 'REQUEST_INITIATED',
      url: 'https://example.com/baz'
    }
  ];
  const completedRequestTrackers: RequestTrackerMessage[] = [
    {
      id: '2',
      message: 'REQUEST_COMPLETED',
      url: 'https://example.com/bar'
    },
    {
      id: '1',
      message: 'REQUEST_COMPLETED',
      url: 'https://example.com/foo'
    },
    {
      id: '3',
      message: 'REQUEST_COMPLETED',
      url: 'https://example.com/baz',
      error: 'ERROR: Baz'
    }
  ];
  let container;

  const renderLoadingWidget = () => {
    store.dispatch(analystActions.trackPendingRequests(pendingRequestTrackers));
    store.dispatch(analystActions.trackCompletedRequests(completedRequestTrackers));
    container = render(
      <Provider store={store}>
        <LoadingWidget />
      </Provider>
    ).container;
  };
  const cleanupLoadingWidget = () => {
    store.dispatch(analystActions.resetRequestTracker());
    cleanup();
  };
  describe('StatusBar widget', () => {
    beforeEach(renderLoadingWidget);
    afterEach(cleanupLoadingWidget);
    it('matches a snapshot', () => {
      expect(container).toMatchSnapshot();
    });
    it('has the loading counter', () => {
      const loadingCounter = container.querySelector('.loading-counter');
      expect(loadingCounter).toMatchSnapshot();
      expect(loadingCounter.innerHTML).toBe(
        `${completedRequestTrackers.length}/${pendingRequestTrackers.length}`
      );
    });
    it('Shows "Completed" when loading finishes', () => {
      cleanup();
      store.dispatch(
        analystActions.trackCompletedRequests([
          {
            id: '4',
            message: 'REQUEST_COMPLETED',
            url: 'https://example.com/foo'
          },
          {
            id: '5',
            message: 'REQUEST_COMPLETED',
            url: 'https://example.com/bar'
          },
          {
            id: '6',
            message: 'REQUEST_COMPLETED',
            url: 'https://example.com/baz',
            error: 'Test Error: Baz'
          }
        ])
      );
      container = render(
        <Provider store={store}>
          <LoadingWidget />
        </Provider>
      ).container;
      expect(screen.getByText('Complete')).toBeInTheDocument();
    });
  });
  describe('LoadingInfoTooltip', () => {
    beforeEach(renderLoadingWidget);
    afterEach(cleanupLoadingWidget);
    it('creates the tooltip on click', () => {
      const button = screen.getByRole('button');
      fireEvent.click(button);
      expect(screen.getByText('Error:')).toBeInTheDocument();
      expect(screen.getByText('ERROR: Baz')).toBeInTheDocument();
      expect(screen.getAllByText('/foo')).toHaveLength(2);
      expect(screen.getAllByText('/bar')).toHaveLength(2);
      expect(screen.getAllByText('/baz')).toHaveLength(3);
    });
  });
});
