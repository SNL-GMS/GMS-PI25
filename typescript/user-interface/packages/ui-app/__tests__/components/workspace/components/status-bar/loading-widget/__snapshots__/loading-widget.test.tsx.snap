// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoadingWidget StatusBar widget has the loading counter 1`] = `
<span
  class="loading-counter"
>
  3/6
</span>
`;

exports[`LoadingWidget StatusBar widget matches a snapshot 1`] = `
<div>
  <span
    aria-haspopup="true"
    class="bp4-popover2-target"
  >
    <div
      class="loading-widget"
      role="button"
    >
      <div
        aria-valuemax="100"
        aria-valuemin="0"
        class="bp4-spinner loading-widget__spinner"
        role="progressbar"
        size="20"
      >
        <div
          class="bp4-spinner-animation"
        >
          <svg
            height="20"
            stroke-width="16.00"
            viewBox="-3.00 -3.00 106.00 106.00"
            width="20"
          >
            <path
              class="bp4-spinner-track"
              d="M 50,50 m 0,-45 a 45,45 0 1 1 0,90 a 45,45 0 1 1 0,-90"
            />
            <path
              class="bp4-spinner-head"
              d="M 50,50 m 0,-45 a 45,45 0 1 1 0,90 a 45,45 0 1 1 0,-90"
              pathLength="280"
              stroke-dasharray="280 280"
              stroke-dashoffset="210"
            />
          </svg>
        </div>
      </div>
      <span
        class="loading-counter"
      >
        3/6
      </span>
      <div
        class="loading-info"
      >
        <span
          class="loading-info__label"
        >
          Loading: 
        </span>
        /baz
      </div>
    </div>
  </span>
</div>
`;
